articles/service-fabric/service-fabric-work-with-reliable-collections.md

```diff
@@ -213,10 +213,53 @@ Internally, Reliable Collections serialize your objects using .NET's DataContrac
+    - performs all operations on the original, V1 collection using the V1 data contracts.
+2. Upgrade service to a new version that
+    - [creates a new, V2 collection](/dotnet/api/microsoft.servicefabric.data.ireliablestatemanager.getoraddasync);
+    - performs each add, update and delete operation on first V1 and then V2 collections in a single transaction;
+    - performs read operations on the V1 collection only.
+3. Copy all data from the V1 collection to the V2 collection.
+    - This can be done in a background process by the service version deployed in step 2.
+    - [Retreieve all keys](/dotnet/api/microsoft.servicefabric.data.collections.ireliabledictionary2-2.createkeyenumerableasync)
+      from the V1 collection. Enumeration is performed with the
+      [IsolationLevel.Snapshot](/dotnet/api/microsoft.servicefabric.data.beta.isolationlevel)
+      by default to avoid locking the collection for the duration of the operation.
+    - For each key, use a separate transaction to
+        - [TryGetValueAsync](/dotnet/api/microsoft.servicefabric.data.collections.ireliabledictionary-2.trygetvalueasync)
+          from the V1 collection.
+        - If the value has already been removed from the V1 collection since the copy process started,
+          the key should be skipped and not resurected in the V2 collection.
+        - [TryAddAsync](/dotnet/api/microsoft.servicefabric.data.collections.ireliabledictionary.tryaddasync)
+          the value to the V2 collection.
+        - If the value has already been added to the V2 collection since the copy process started,
+          the key should be skipped.
+        - The transaction should be committed only if the `TryAddAsync` returns `true`.
+        - Value access APIs use the [IsolationLevel.ReadRepeatable](/dotnet/api/microsoft.servicefabric.data.beta.isolationlevel)
+          by default and rely on locking to guarantee that the values aren't modified by another caller until the transaction is committed or aborted.
+4. Upgrade service to a new version that
+    - performs read operations on the V2 collection only;
+    - still performs each add, update and delete operation on first V1 and then V2 collections to maintain the option of rolling back to V1.
+5. Comprehensively test the service and confirm it is working as expected.
+    - If you missed any value access operation that wasn't updated to work on both V1 and V2 collection, you may notice missing data.
+    - If any data is missing roll back to Step 1, remove the V2 collection and repeat the process.
+6. Upgrade service to a new version that
+    - performs all operations on the V2 collection only;
+    - going back to V1 is no longer possible with a service rollback and would require rolling forward with reversed steps 2-4.
+7. Upgrade service a new version that
+    - [removes the V1 collection](/dotnet/api/microsoft.servicefabric.data.ireliablestatemanager.removeasync).
+8. Wait for log truncation.

```

Comment:
==repost since the old one got outdated without answer.==

should SF still block upgrade if the log is the only one that requires old schema?

Reply:
Unfortunately, it can't. Any data stored in the logs with old data contracts before they are checkpointed still needs them to be loaded by replicas from disk or replicated.

Reply:
if the reliable collection is gone already, what is the purpose to keep these logs, in another question: when SF processes the logs, should it check if the corresponding reliable collection exists first?